app/services/badges/award_community_wellness.rb

```diff
@@ -0,0 +1,80 @@
+module Badges
+  class AwardCommunityWellness
+    REWARD_STREAK_WEEKS = [1, 2, 4, 8, 16, 32].freeze
+
+    def self.call
+      new.call
+    end
+
+    def call
+      # These are the users 'eligible' to be awarded the badge, which attempts
+      # to make the subsequent queries/iterations more performant by using a
+      # subset of users and not large joins.
+      users = User.where(id: multiple_comment_user_ids)
+
+      users.find_each do |user|
+        week_streak = 0
+        (1..REWARD_STREAK_WEEKS.last).each do |week|
+          break unless wellness_goal_x_weeks_ago?(user, week)
+
+          week_streak += 1
+        end
```

Comment:
This is a really smart optimization, and a fantastic explanation!

Reply:
Thanks @joshpuetz! A condensed SQL query will be more performant in comparison though. Thanks to the help from @jeremyf and @jgaskins to write it I was just able to push an updated approach (still has some pending details to polish).

Greedy algorithms are fun, so maybe a solution like this will be viable in some other time/area of the codebase 😄